crates/evm-prover/src/prover/programs/message.rs

```diff
@@ -0,0 +1,348 @@
+//! The prover implementation of the Hyperlane Message circuit proves all messages that have occurred in between
+//! two given heights against a given EVM block height.
+
+#![allow(dead_code)]
+use std::{
+    str::FromStr,
+    sync::{Arc, RwLock},
+    time::Duration,
+};
+
+use alloy_primitives::{hex::FromHex, Address, FixedBytes};
+use alloy_provider::{fillers::FillProvider, Provider, ProviderBuilder, WsConnect};
+use alloy_rpc_types::{EIP1186AccountProofResponse, Filter};
+use anyhow::Result;
+use evm_hyperlane_types_sp1::{HyperlaneMessageInputs, HyperlaneMessageOutputs};
+use evm_state_queries::hyperlane::indexer::HyperlaneIndexer;
+use evm_state_types::events::Dispatch;
+use evm_storage_proofs::{
+    client::EvmClient,
+    types::{HyperlaneBranchProof, HyperlaneBranchProofInputs, HYPERLANE_MERKLE_TREE_KEYS},
+};
+use reqwest::Url;
+use sp1_sdk::{include_elf, EnvProver, ProverClient, SP1ProofMode, SP1ProofWithPublicValues, SP1Stdin};
+use storage::hyperlane::{message::HyperlaneMessageStore, snapshot::HyperlaneSnapshotStore};
+use tokio::time::sleep;
+
+use crate::prover::{ProgramProver, ProverConfig};
+
+pub type DefaultProvider = FillProvider<
+    alloy_provider::fillers::JoinFill<
+        alloy_provider::Identity,
+        alloy_provider::fillers::JoinFill<
+            alloy_provider::fillers::GasFiller,
+            alloy_provider::fillers::JoinFill<
+                alloy_provider::fillers::BlobGasFiller,
+                alloy_provider::fillers::JoinFill<
+                    alloy_provider::fillers::NonceFiller,
+                    alloy_provider::fillers::ChainIdFiller,
+                >,
+            >,
+        >,
+    >,
+    alloy_provider::RootProvider,
+>;
+
+const TIMEOUT: u64 = 6; // in seconds
+const DISTANCE_TO_HEAD: u64 = 32; // in blocks
+
+/// The ELF (executable and linkable format) file for the Succinct RISC-V zkVM.
+pub const EVM_HYPERLANE_ELF: &[u8] = include_elf!("evm-hyperlane-program");
+
+/// AppContext encapsulates the full set of RPC endpoints and configuration
+/// needed to fetch input data for execution and data availability proofs.
+///
+/// This separates RPC concerns from the proving logic, allowing `AppContext`
+/// to be responsible for gathering the data required for the proof system inputs.
+pub struct AppContext {
+    pub celestia_rpc: String,
+    // reth http, for example http://127.0.0.1:8545
+    pub evm_rpc: String,
+    // reth websocket, for example ws://127.0.0.1:8546
+    pub evm_ws: String,
+    pub mailbox_address: Address,
+    pub merkle_tree_address: Address,
+    pub trusted_state: RwLock<TrustedState>,
+}
+
+/// TrustedState encapsulates the trusted state of the prover
+pub struct TrustedState {
+    // the index of the snapshot that we will load from the db, initially 0 (empty by default)
+    snapshot_index: u64,
+    // the index of the last message we proofed successfully, initially 0
+    height: u64,
+}
```

Comment:
This is a little bit confusing to me for some reason. Maybe could be renamed from TrustedState to better reflect its purpose. 

Are we not storing the tree branches/snapshots keyed by height?